waspc/waspls/test/Wasp/LSP/CompletionTest.hs

```diff
@@ -0,0 +1,145 @@
+import qualified Language.LSP.Types.Lens as LSP
+import System.FilePath (replaceExtension, takeBaseName)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.Golden (findByExtension, goldenVsStringDiff)
+import Wasp.Analyzer.Parser.ConcreteParser (parseCST)
+import qualified Wasp.Analyzer.Parser.Lexer as Lexer
+import Wasp.LSP.Completion (getCompletionsAtPosition)
+import Wasp.LSP.ServerState (ServerState (ServerState, _cst, _currentWaspSource, _latestDiagnostics))
+
+-- | Run test cases in ./completionTests directory
+--
+-- See 'readCompletionTest' for the format of a test case.
+test_CompletionLists :: IO TestTree
+test_CompletionLists = do
+  inputFiles <- findByExtension [".wasp"] "./waspls/test/Wasp/LSP/completionTests"
+  return $
+    testGroup "Wasp.LSP.Completion.getCompletionsAtPosition" $
+      makeCompletionTest <$> inputFiles
+
+-- | Takes a completion test case and produces the list of completion items
+runCompletionTest :: String -> String
+runCompletionTest source =
+  let (code, position) = readCompletionTest source
+      tokens = Lexer.lex code
+      parsedCST = snd $ parseCST tokens
+      serverState =
+        ServerState
+          { _currentWaspSource = code,
+            _latestDiagnostics = [],
+            _cst = Just parsedCST
+          }
+      (completionItems, _log) = runLog $ evalStateT (getCompletionsAtPosition position) serverState
+      fmtedCompletionItems = map fmtCompletionItem completionItems
+
+      fmtCompletionItem :: LSP.CompletionItem -> String
+      fmtCompletionItem item =
+        concat
+          [ "  label={",
+            show (item ^. LSP.label),
+            "} kind={",
+            show (item ^. LSP.kind),
+            "} detail={",
+            show (item ^. LSP.detail),
+            "}"
+          ]
+   in "Completion Items:\n" ++ unlines fmtedCompletionItems
+
+-- | Parses a completion test case into a pair of the wasp source code to
+-- run completion on and the position to get the completion list at.
+--
+-- = Format
+--
+-- A normal wasp file, but with two addition
+-- 1) Begins with "//! test/completion" (this is future-proofing in case we
+--    add editor support for these test files).
+-- 2) In one spot in the file, add a "completion marker", which is a "|" with
+-- a "^" at the same column on the line below. The line containing "^" should be
+-- blank except for whitespace before the "^" and the "^" itself.
+--
+-- The "|", "^", and the extra line are not part of the wasp source code and
+-- are not included in the returned code. The preamble comment is included in
+-- the returned code.
+--
+-- If there is more than one completion marker in the input, only the first
+-- marker is recognized, and the rest are left in the source code.
+--
+-- === __Example__
+--
+-- @
+-- //! test/completion
+-- app todoApp {
+--   |
+--   ^
+-- }
+-- @
+--
+-- This test case checks completions after the 2 spaces on the 2nd line.
+readCompletionTest :: String -> (String, LSP.Position)
```

Comment:
How about `parseCompletionTestInput`? Or `parseWaspSourceWithCursor`, if you go with that name.
`parse`, because it actually returns new things -> new source without the cursor, and cursor position. I would say that is proper parsing, not just "reading" something.